### DISCUSSION TITLE
Perform error: rpc error: code = ResourceExhausted desc = No free threads in thread pool

### DISCUSSION THREAD

**Discussion by mantouri** [anycable/anycable#180] [Q&A]
i use custom setting when start anycable server to try speed up server
```bundle exec anycable --log-file=log/anycable.log --log-level debug --server-command "\anycable-go --host=0.0.0.0 --port=59901 --rpc_concurrency=1020 --hub_gopool_size=640"```
but i will show error like that 
<img width="1130" alt="截圖 2023-01-03 下午2 42 22" src="https://user-images.githubusercontent.com/40049356/210310529-76b617e3-9c61-47f8-a009-44e0d97b2a37.png">
how to fix it ? 
and what is correctly config setting if I want support 10k websocket client request, every response can in 6 second to finish.

**Comment by palkan** [comment]
You need to tune concurrency at both ends. Right now, your RPC server has a default pool size (I guess) of 30 threads, but you try to make ~1k concurrent requests from the `anycable-go` side. AnyCable-Go concurrency is not a number of concurrent connections, but a number of concurrent RPC requests. The best way to pick the right values is to monitor the actual usage, see https://docs.anycable.io/anycable-go/instrumentation?id=metrics-and-what-can-we-learn-from-them.

> if I want support 10k websocket client request, every response can in 6 second to finish.

It depends on what these clients are doing; do you have any monitoring set up?

**Reply by reguchibr** [reply]
Hi @palkan, I've been facing the same error.

I've configured the following:

1) anycable-go deployment
ANYCABLE_RPC_CONCURRENCY = 100

2) anycable-rpc deployment
ANYCABLE_RPC_POOL_SIZE = 220
ANYCABLE_RPC_SERVER_ARGS__MAX_CONNECTION_AGE_MS = 300000
MAX_THREADS = 230

Everything works fine for a few hours but then there are huge spikes on rpc call I guess due to the "ResourceExhausted" errors I see on logs.

I'm running anycable on AWS EKS 1.23.

I've six pods for anycable-go and 8 pods for anycable-rpc.

Any help is appreciated.

Thanks,
Roger

**Reply by palkan** [reply]
> I've six pods for anycable-go and 8 pods for anycable-rpc.

How do you load balance the traffic?

Any deployments happened around the time you see the spike? What happened in 10-15 minutes after it?

**Reply by reguchibr** [reply]
Nginx + headless service

There have been no deployments around the spikes. The connections are stable 10-15 minutes after deploying.

<img width="328" alt="Screenshot 2023-02-20 at 22 18 28" src="https://user-images.githubusercontent.com/33463204/220239008-67d80c0f-63d8-4db3-ad76-fcccb1b30c07.png">

**Reply by palkan** [reply]
Thanks!

I should have been more precise: how do you distribute load across RPC services? Also NGINX (with grpc module) or something else? What about health checks (if it's Nginx)?

With your configuration (concurrency settings), the ResourceExhausted errors can only happen if RPC calls are not distributed uniformly (so, some instances receive mode calls than they can handle). The overall concurrency _capacity_ at the anycable-go side is 600, so, if for some reason at some moment in time only 2/8 RPC instances were live, you can make them exhausted.

**Reply by mantouri** [reply]
> You need to tune concurrency at both ends. Right now, your RPC server has a default pool size (I guess) of 30 threads, but you try to make ~1k concurrent requests from the `anycable-go` side. AnyCable-Go concurrency is not a number of concurrent connections, but a number of concurrent RPC requests. The best way to pick the right values is to monitor the actual usage, see https://docs.anycable.io/anycable-go/instrumentation?id=metrics-and-what-can-we-learn-from-them.
> 
> > if I want support 10k websocket client request, every response can in 6 second to finish.
> 
> It depends on what these clients are doing; do you have any monitoring set up?

Yes, we monitor anycable metrics set up on grafana , and daily average 
Go routines number
<img width="335" alt="截圖 2023-03-22 下午11 23 18" src="https://user-images.githubusercontent.com/40049356/226953277-2db9b4fb-b5eb-48b5-9da8-2590a81577c0.png">
RPC calls
<img width="306" alt="截圖 2023-03-22 下午11 23 09" src="https://user-images.githubusercontent.com/40049356/226953283-5efc3300-ca70-4a6a-824d-2fbd361d7887.png">

adjust the conditional data below, speed will be improve ?
`--read_buffer_size=4096 --write_buffer_size=4096 --max_message_size=512000 --rpc_concurrency=100 
`

**Reply by palkan** [reply]
Could you please provide a legend for the RPC calls chart? Are these different hosts or different metrics?

**Reply by mantouri** [reply]
different host collection on one chart , like this on pic show 
<img width="183" alt="截圖 2023-03-23 上午10 14 31" src="https://user-images.githubusercontent.com/40049356/227082224-aa6ddca0-d557-4c3d-a0d5-12d086d5effd.png">

**Comment by reguchibr** [comment]
Hi @palkan,

We are using client-side load balancing via ANYCABLE_RPC_HOST. This env var points to a headless service.

About health checks, do you mean http-health-port and readinessProbe config ?

With my config, the ResourceExhausted errors are due to the headless service not working or rpc pods getting down, is that it?

Thanks

**Reply by palkan** [reply]
> the headless service not working or rpc pods getting down, is that it?

Yeah, looks so.

> We are using client-side load balancing via ANYCABLE_RPC_HOST

One suggestion is to try our edge version (release is coming later this week; hopefully, tomorrow) which adds the `grpc_active_conn_num` metrics, showing how many gRPC connections are currently active (should be one per RPC endpoint in case of client-side load-balancing). This way, you can see if your WS servers connected to all RPC servers (and, thus, distribute calls across all of them) or not.
